### PR TITLE
Simplify bulk request execution

### PR DESCRIPTION
**Pull request by dakrone**
This is a bespoke backport of #20109 for 5.x:

Currently, bulk item requests can be any ActionRequest, this PR restricts bulk
item requests to DocumentRequest. This simplifies handling failures during bulk
requests. Additionally, a new enum is added to DocumentRequest to represent the
intended operation to be performed by a document request (create, index, update
and delete), which was previously represented with a mix of strings and index
request operation type.

Now, index request operation type reuses the new enum to specify whether the
request should create or index a document. Restricting bulk requests to
DocumentRequest further simplifies execution of shard-level bulk operations to
use the same failure handling for index, delete and update operations. This PR
also fixes a bug which executed delete operations twice for replica copies while
executing bulk requests.

Relates to #19105 and #20109

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 1]
```diff
@@ -24,6 +24,9 @@
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ActionRequest;
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: we can remove the unused imports here (namely: `ActionRequest`, `DocWriteRequest.OpType`, `ReplicationRequest`, `ShardId` and `RestStatus`)

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -129,203 +126,87 @@ protected boolean resolveIndex() {
...
+            } else {
+                logger.debug((Supplier<?>) () -> new ParameterizedMessage("{} failed to execute bulk item ({}) {}",
+                                request.shardId(), docWriteRequest.opType().getLowercase(), request), e);
+            }
             // if its a conflict failure, and we already executed the request on a primary (and we execute it
             // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
             // then just use the response we got from the successful execution
             if (item.getPrimaryResponse() != null && isConflictException(e)) {
-                setResponse(item, item.getPrimaryResponse());
+                 setResponse(item, item.getPrimaryResponse());
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: minor: fix indentation

**Review by areek — APPROVED** [review]
Thanks @dakrone! LGTM, left a couple of minor comments

**Comment by dakrone** [comment]
Thanks for taking a look @areek!

**Pull request merged**